src/compiler/checker.ts

```diff
@@ -28040,6 +28036,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getUnionType(map(clauseWitnesses, text => text ? narrowTypeByTypeName(type, text) : neverType));
         }
 
+        function narrowTypeBySwitchOnTrue(type: Type, switchStatement: SwitchStatement, clauseStart: number, clauseEnd: number): Type {
+            // Equal start and end denotes implicit fallthrough; undefined marks explicit default clause.
+            const defaultIndex = findIndex(switchStatement.caseBlock.clauses, clause => clause.kind === SyntaxKind.DefaultClause);
+            const hasDefaultClause = clauseStart === clauseEnd || (defaultIndex >= clauseStart && defaultIndex < clauseEnd);
+            if (hasDefaultClause) {
+                // Here we'd need to find the opposite of all of the other cases, but without control flow,

```

Comment:
It's not just the opposite because you can fall through to a default. For example

```ts
declare const f: 'a' | 'b' | 'c';

switch (true) {
  case f === "a":
  default:
    f;
  case f === "b":
}
```

Ideally we would know that `f` is `"a" | "c"`.